tests/Unit/Domain/BoundaryConditions/Test_BoundaryCondition.cpp

```diff
@@ -0,0 +1,51 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <charm++.h>
+#include <memory>
+#include <pup.h>
+#include <string>
+
+#include "Domain/BoundaryConditions/BoundaryCondition.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "Parallel/CharmPupable.hpp"
+
+namespace {
+class BoundaryCondition
+    : public domain::boundary_conditions::BoundaryCondition {
+ public:
+  BoundaryCondition() = default;
+
+  std::unique_ptr<domain::boundary_conditions::BoundaryCondition> get_clone()
+      const noexcept override {
+    return std::make_unique<BoundaryCondition>(*this);
+  }
+
+  explicit BoundaryCondition(CkMigrateMessage* msg) noexcept
+      : domain::boundary_conditions::BoundaryCondition(msg) {}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+  WRAPPED_PUPable_decl_base_template(
```

Comment:
Out of curiosity, did you also run into https://github.com/UIUC-PPL/charm/issues/3235? Do you have any insight into that issue; would you say it's a bug or a feature?

Reply:
No

Reply:
well, according to the docs (ours and charm's) this should just be `WRAPPED_PUPable_decl` (no `_template`)

Reply:
You can use the template one in either case. I don't really see a reason to change it. Probably the template one should always be used because it's more correct. I forget the edge case but Charm++ used to not have this supported and we had to patch Charm++ to get it to work, I think.

Reply:
yea don't change it since then you'll get linker errors :) I think this is either a bug in charm or in their docs.

Reply:
I wonder if we should just remove the non-template wrapper and tell people to use the template one because it works :P

Reply:
I learned from @wthrowe today that the `WRAPPED_PUPable_decl` macro has to be matched with a `PUPable_def` at file scope. I have actually spent way too much time fighting with these macros to make the compiler, linker and runtime happy, so I think it would be good if we could make this easier somehow. Especially because these macros have to be used by new contributors to the code who just want to add e.g. a boundary condition or an analytic solution. So anything we can do to make them easier to use would be a gain I think.

Reply:
If the `_template` one works generally, I think that's the way to go. I believe that one defines the function inline because class templates make the whole thing even more difficult